File: ports/pyagentic/pyagentic/stream.py
"""The stream substrate — an agent as a materialized view over a stream of events.

The portable counterpart of a Flink source / Kafka consumer / Pekko Source. A
``Channel`` is a pull-based event source; the ``StreamRuntime`` drains it through a
backend ``Runtime`` as agent turns, so ``Runtime.submit(event)`` is just the one-shot
sugar and this is the streaming form. ``EventObserver`` callables see every event
before it becomes a turn — the seam CEP matchers, window aggregators, and tracers
plug into in later phases without the agent graph needing to know about them.

Per-conversation ordering stays the runtime's concern (the ``LocalRuntime`` per-key
lock, a Kafka partition, a Ray actor); the stream loop just preserves arrival order.
"""

from __future__ import annotations

from collections import deque
from threading import Lock
from typing import TYPE_CHECKING, Callable, Generic, Iterable, List, Optional, Protocol, TypeVar

from .core import Event, TurnResult
from .runtime import Runtime
from .trace import NoopTracer, Tracer

if TYPE_CHECKING:
    from .timers import TimerService

T = TypeVar("T")


class Channel(Protocol[T]):
    """A source of events. Pull-based: :meth:`poll` returns the next available event,
    or ``None`` when none is available right now. Never blocks. Bounded sources (a seed
    list) eventually return ``None`` forever; unbounded sources (a queue) may return
    ``None`` transiently and more later."""

    def poll(self) -> Optional[T]:
        """The next event if one is available, else ``None``. Never blocks."""
        ...


class SeedChannel(Generic[T]):
    """A bounded channel that replays a fixed list of events in order, then is empty
    (``None``) forever."""

    def __init__(self, events: Iterable[T]) -> None:
        self._it = iter(list(events))

    @classmethod
    def of(cls, *events: T) -> "SeedChannel[T]":
        return cls(events)

    def poll(self) -> Optional[T]:
        return next(self._it, None)


class QueueChannel(Generic[T]):
    """An unbounded in-memory channel: producers :meth:`offer` events; the runtime
    polls them in FIFO order. Thread-safe, so a producer thread and the stream loop can
    share it."""

    def __init__(self) -> None:
        self._queue: "deque[T]" = deque()
        self._guard = Lock()

    def offer(self, event: T) -> "QueueChannel[T]":
        with self._guard:
            self._queue.append(event)
        return self

    def poll(self) -> Optional[T]:
        with self._guard:
            return self._queue.popleft() if self._queue else None


# An observer sees every event the StreamRuntime drives, before it becomes a turn.
EventObserver = Callable[[Event], None]


class StreamRuntime:
    """Drives a :class:`Channel` of events through a backend :class:`Runtime` as agent
    turns — the realization of the thesis that an agent is a materialized view over a
    stream of events. ``Runtime.submit(event)`` is the one-shot sugar; this is the
    streaming form."""

    def __init__(self, runtime: Runtime) -> None:
        self._runtime = runtime
        self._observers: List[EventObserver] = []
        self._tracer: Tracer = NoopTracer()

    def observe(self, observer: EventObserver) -> "StreamRuntime":
        """Register an observer of the raw event stream (chainable)."""
        self._observers.append(observer)
        return self

    def with_tracer(self, tracer: Optional[Tracer]) -> "StreamRuntime":
        """Trace turns and timer fires through this tracer (chainable; default no-op)."""
        self._tracer = tracer if tracer is not None else NoopTracer()
        return self

    def _submit_traced(self, span_name: str, event: Event) -> TurnResult:
        """Notify observers, then submit ``event`` as a turn inside a trace span. The span
        carries the conversation id, chosen path, and ok flag (``"true"``/``"false"`` to
        mirror Java's ``Boolean.toString``), and one ``tool:<id>`` event per tool call."""
        for observer in self._observers:
            observer(event)
        span = self._tracer.start(span_name)
        span.attr("conversation", event.conversation_id)
        r = self._runtime.submit(event)
        span.attr("path", r.path).attr("ok", str(r.ok).lower())
        for tc in r.tool_calls:
            span.event("tool:" + tc)
        span.end()
        return r

    def run(self, channel: Channel[Event]) -> List[TurnResult]:
        """Drain every currently-available event from the channel as a turn, in arrival
        order, and return the results. Observers see each event first. Returns when the
        channel next reports empty (``poll() is None``)."""
        results: List[TurnResult] = []
        while True:
            event = channel.poll()
            if event is None:
                break
            results.append(self._submit_traced("turn", event))
        return results

    def fire_due_timers(self, timer_service: "TimerService", now: int) -> List[TurnResult]:
        """Advance ``timer_service`` to ``now`` and fire each due timer's payload through
        the backend runtime as a turn, in deadline order. Observers see each payload first,
        exactly as with a channel event — a fired timer is just another event entering the
        stream. Returns the results in deadline (then schedule) order."""
        results: List[TurnResult] = []
        for timer in timer_service.advance_to(now):
            results.append(self._submit_traced("timer.fire", timer.payload))
        return results

File: agentic-clj/src/agentic/stream.clj
(ns agentic.stream
  "Phase 1: the stream substrate — the portable realization of the project's thesis that an agent is a
   materialized view over a stream of events. A Channel is a pull-based event source; a stream-runtime
   drives a channel of events through agentic.core/submit as agent turns, in arrival order. Observers
   see each raw event before it becomes a turn — the seam CEP matchers, window aggregators and tracers
   plug into in later phases. Mirror of jagentic-core's org.jagentic.core.stream.*."
  (:require [agentic.core :as core]
            [agentic.timers :as timers]))

;; ---- Channel: a pull-based event source ----

(defprotocol Channel
  "A source of events — the portable counterpart of a Flink source / Kafka consumer / Pekko Source.
   Pull-based: (poll) returns the next available event, or nil when none is available right now. Never
   blocks. Bounded sources (a seed list) eventually return nil forever; unbounded sources (a queue) may
   return nil transiently and more later."
  (poll [_] "The next event if one is available, else nil. Never blocks."))

;; ---- seed-channel: bounded, replays a fixed seq then nil forever ----

(deftype SeedChannel [!remaining]
  Channel
  (poll [_]
    ;; Atomic pop-front: swap! returns the new state, so re-derive the popped head from the old state.
    (let [old (first (swap-vals! !remaining next))]
      (first old))))

(defn seed-channel
  "A bounded channel that replays the given events in order, then is nil forever."
  [events]
  (->SeedChannel (atom (seq events))))

;; ---- queue-channel: unbounded thread-safe FIFO ----

(deftype QueueChannel [!queue]
  Channel
  (poll [_]
    ;; Atomically pop the head: swap-vals! gives [old new]; the head is (peek old).
    (peek (first (swap-vals! !queue pop)))))

(defn queue-channel
  "An unbounded, thread-safe in-memory channel: producers (offer) events; the runtime polls them in
   FIFO order. A producer thread and the stream loop can share it safely."
  []
  (->QueueChannel (atom clojure.lang.PersistentQueue/EMPTY)))

(defn offer
  "Enqueue an event on a queue-channel; returns the channel so calls chain."
  [ch event]
  (swap! (.-!queue ^QueueChannel ch) conj event)
  ch)

;; ---- stream-runtime: drive a channel of events as agent turns ----

(defn stream-runtime
  "Build a stream-runtime over a `system` (what agentic.core/submit takes). Modelled as plain data —
   {:system :observers} — so observe/run are ordinary functions."
  [system]
  {:system system :observers []})

(defn observe
  "Register an observer of the raw event stream. An observer is a fn (fn [event] ...) called for each
   event before it becomes a turn. Returns the updated stream-runtime so calls chain."
  [sr observer-fn]
  (update sr :observers conj observer-fn))

(defn run
  "Drain every currently-available event from the channel as a turn, in arrival order, and return the
   vector of turn-results. Each observer sees the event first, then (core/submit system event) runs.
   Returns when the channel next reports nil (empty)."
  [{:keys [system observers]} channel]
  (loop [results (transient [])]
    (if-let [event (poll channel)]
      (do
        (doseq [obs observers] (obs event))
        (recur (conj! results (core/submit system event))))
      (persistent! results))))

(defn fire-due-timers
  "Fire every timer due at logical time `now` through the runtime, in deadline order. Pulls the due
   timers from `timer-service` via agentic.timers/advance-to (ascending by :fire-at, schedule order as
   the stable tie-break), and for each — exactly like (run) — lets observers see the payload event
   first, then submits it as a turn. Returns the vector of turn-results in deadline order. This is the
   seam SLAs / escalate-after-N / scheduled follow-ups fire on. `timer-service` must satisfy the
   agentic.timers/TimerService protocol."
  [{:keys [system observers]} timer-service now]
  (let [due (timers/advance-to timer-service now)]
    (mapv (fn [{:keys [payload]}]
            (doseq [obs observers] (obs payload))
            (core/submit system payload))
          due)))
